### PR TITLE
Drop-down menus

### PR DESCRIPTION
**Pull request by NeoAdonis**
Part of #250; supersedes #257. See `TestCaseDropDown` visual test for example use-case.

RFC; I want to know if more functionality is needed, or if there's need to simplify classes more.

Current functionality:
- Adding items
- Selecting items
- Bi-directional binding
- Clearing item list
- Customize style and animations

### CODE REVIEW

**Comment by SirCmpwn** [comment]
How about dropping the generics and making each DropDownMenuItem a Container that just displays its children? For headers it could have a Clickable property.

**Comment by SirCmpwn** [comment]
Why does StyledDropDownMenu have to be different from DropDownMenu?

**Comment by peppy** [comment]
the generics are only used for associated items, though? i think this is fine, instead of winforms for instance (when you get an `object` back in such cases)

**Comment by SirCmpwn** [comment]
Yeah, I take that back. I didn't realize that those classes were defined in the VisualTests, I thought they were part of framework at first.

**Comment by SirCmpwn** [comment]
Overall pretty good. :+1:. I'd just change the label to be a container.

File: osu.Framework/Graphics/UserInterface/DropDownMenuItem.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        {
+            if (!IsLoaded)
+                return;
+
+            FormatCaret();
+        }
+
+        protected virtual void FormatCaret()
+        {
+            (Caret as SpriteText).Text = IsSelected ? @">>" : @">";
```
Comment (NeoAdonis) [inline osu.Framework/Graphics/UserInterface/DropDownMenuItem.cs]: FWIW I wanted to change label font weight (bold when selected). Did this instead as I think the former is not possible yet, but yeah: what you propose is better than what I tried to do. :P

**Comment by peppy** [comment]
I've only briefly gone over the code, but here's a few pieces of feedback from visual inspection:

- transition doesn't look the same on the second revealing of a dropdown (appears instantly rather than moving to correct place.



- there's no background behind elements after they are revealed. you can see things behind the dropdown which looks wrong.



- corners should probably be rounded in this example

image

**Comment by NeoAdonis** [comment]
Thanks for pointing out that transition. :P About second one, maybe putting a semi-transparent (or opaque) background would work? What do you think? That or disallowing scrolling out of bounds. And yeah, one could round corners; I'll add that.

**Comment by peppy** [comment]
adding a background layer is a good solution, yes.

**Pull request merged**